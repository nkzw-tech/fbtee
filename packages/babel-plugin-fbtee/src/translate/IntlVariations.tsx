import invariant from 'invariant';

export const IntlNumberVariations = {
  // 0b01000
  FEW: 20,
  // 0b10100
  MANY: 12,
  // 0b10000
  ONE: 4,
  // 0b01100
  OTHER: 24,
  // 0b00100
  TWO: 8,
  // Cast below values to IntlVariationsEnum
  ZERO: 16, // 0b11000
} as const;

export type IntlVariations =
  (typeof IntlNumberVariations)[keyof typeof IntlNumberVariations];

// Must match with `IntlVariations.js`
export const Gender = {
  FEMALE: 2,
  // Cast below values to IntlVariationsEnum
  MALE: 1,
  UNKNOWN: 3,
} as const;

export type Gender = (typeof Gender)[keyof typeof Gender];

// Two bitmasks for representing gender/number variations.  Give a bit
// between number/gender in case CLDR ever exceeds 7 options
export const Mask = {
  GENDER: 3,
  NUMBER: 28,
} as const;

export type IntlVariationMaskValue = (typeof Mask)[keyof typeof Mask];

export const FbtVariationType = {
  GENDER: 1,
  NUMBER: 2,
  PRONOUN: 3,
} as const;

export type IntlFbtVariationTypeValue =
  (typeof FbtVariationType)[keyof typeof FbtVariationType];

export const VIEWING_USER = '__viewing_user__';
export const SUBJECT = '__subject__';

// Gender variation key used in JSFBT to represent any gender
export const GENDER_ANY = '*' as const;
// Number variation key used in JSFBT to represent "many" (i.e. non-exactly one)
export const NUMBER_ANY = '*' as const;

// This is not CLDR, but an fbt-specific marker that exists so that
// singular phrases are not overwritten by multiplexed plural phrases
// with a singular entry
export const EXACTLY_ONE = '_1' as const;

const SPECIALS = {
  // The default entry.  When no entry exists, we fallback to this in the fbt
  // table access logic.
  '*': true,
  EXACTLY_ONE: true,
} as const;

export function getType(n: IntlVariationMaskValue): IntlVariationMaskValue {
  invariant(isValidValue(n), 'Invalid NumberType: %s', n);

  return n & Mask.NUMBER ? Mask.NUMBER : Mask.GENDER;
}

export function isValidValue(value: string | number): boolean {
  const num = Number(value);
  return (
    SPECIALS[value as keyof typeof SPECIALS] ||
    (num & Mask.NUMBER && !(num & ~Mask.NUMBER)) ||
    (num & Mask.GENDER && !(num & ~Mask.GENDER))
  );
}
