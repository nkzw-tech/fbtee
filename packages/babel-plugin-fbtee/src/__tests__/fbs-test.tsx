import { describe, expect, it } from '@jest/globals';
import TestFbtEnumManifest from '../__mocks__/TestFbtEnumManifest.tsx';
import {
  jsCodeFbtCallSerializer,
  snapshotTransform,
  withFbsImportStatement,
} from './FbtTestUtil.tsx';

expect.addSnapshotSerializer(jsCodeFbtCallSerializer);

describe('Test declarative (jsx) <fbs> syntax translation', () => {
  it('should convert a simple string', () => {
    expect(
      snapshotTransform(
        withFbsImportStatement(
          `const fbsElem = <fbs desc='str_description'>a simple string</fbs>;`,
        ),
      ),
    ).toMatchSnapshot();
  });
  it('should convert a string with a parameter', () => {
    expect(
      snapshotTransform(
        withFbsImportStatement(`
          const fbsElem = <fbs desc='str_description'>
            a string with a
            <fbs:param name="param name">{parameter}</fbs:param>
          </fbs>;
        `),
      ),
    ).toMatchSnapshot();
  });
  it('should convert a common string', () => {
    expect(
      snapshotTransform(
        withFbsImportStatement(`
          const fbsCommonElem = <fbs common={true}>Post</fbs>;
        `),
        {
          fbtCommon: {
            Post: 'Button to post a comment',
          },
        },
      ),
    ).toMatchSnapshot();
  });
  it('should reject an <fbs> child element', () => {
    expect(() =>
      snapshotTransform(
        withFbsImportStatement(`
          const fbsElem = <fbs desc='str_description'>
            a simple string
            <fbs>nested</fbs>
          </fbs>;
        `),
      ),
    ).toThrow(`Don't put <fbs> directly within <fbs>.`);
  });
  it('should reject an <fbt> child element', () => {
    expect(() =>
      snapshotTransform(
        withFbsImportStatement(`
          const fbsElem = <fbs desc='str_description'>
            a simple string
            <fbt>nested</fbt>
          </fbs>;
        `),
      ),
    ).toThrow(`Don't put <fbt> directly within <fbs>.`);
  });
  it('should reject an <fbt:param> child element', () => {
    expect(() =>
      snapshotTransform(
        withFbsImportStatement(`
          const fbsElem = <fbs desc='str_description'>
            a simple string
            <fbt:param name="param name">{parameter}</fbt:param>
          </fbs>;
        `),
      ),
    ).toThrow(`Don't mix <fbt> and <fbs> JSX namespaces.`);
  });
  it('should handle <fbs:enum>', () => {
    expect(
      snapshotTransform(
        withFbsImportStatement(`
          import aEnum from 'Test$FbtEnum';
          var x = (
            <fbs desc="enums!">
              Click to see
              <fbs:enum enum-range={aEnum} value={id} />
            </fbs>
          );
        `),
        { fbtEnumManifest: TestFbtEnumManifest },
      ),
    ).toMatchSnapshot();
  });
});

describe('Test functional fbs() syntax translation', () => {
  it('should convert a simple string', () => {
    expect(
      snapshotTransform(
        withFbsImportStatement(
          `const fbsCall = fbs('a simple string', 'str_description');`,
        ),
      ),
    ).toMatchSnapshot();
  });
  it('should convert a string with a gender parameter', () => {
    expect(
      snapshotTransform(
        withFbsImportStatement(`
          import IntlVariations from 'IntlVariations';
          const fbsCall = fbs(
            'a string with a ' + fbs.param('param name', parameter, {gender: IntlVariations.GENDER_MALE}),
            'str_description'
          );
        `),
      ),
    ).toMatchSnapshot();
  });
  it('should convert a common string', () => {
    expect(
      snapshotTransform(
        withFbsImportStatement(`const fbsCommonCall = fbs.c('Post');`),
      ),
    ).toMatchSnapshot();
  });

  it('should reject an fbt parameter', () => {
    expect(() =>
      snapshotTransform(
        withFbsImportStatement(`
          const fbsCall = fbs(
            'a string with a ' + fbt.param('param name', parameter, {gender: 'male'}),
            'str_description'
          );
        `),
      ),
    ).toThrow(`fbs: unsupported node: CallExpression`);
  });

  it('should throw when using fbs() and the fbs variable is not bound', () => {
    expect(() =>
      snapshotTransform(
        `const fbsCall = fbs(
          'basic',
          'str_description'
        );`,
      ),
    ).toThrow(`fbs is not bound. Did you forget to import fbs?`);
  });

  it('should throw when using <fbs> and the fbs variable is not bound', () => {
    expect(() =>
      snapshotTransform(
        `const fbsCall = <fbs desc="str_description">basic</fbs>;`,
      ),
    ).toThrow(`fbs is not bound. Did you forget to import fbs?`);
  });

  it('should handle fbs.enum', () => {
    expect(
      snapshotTransform(
        withFbsImportStatement(`
          import aEnum from 'Test$FbtEnum';
          var x = fbs('Click to see ' + fbs.enum(id, aEnum), 'enums!');
        `),
        { fbtEnumManifest: TestFbtEnumManifest },
      ),
    ).toMatchSnapshot();
  });
});
