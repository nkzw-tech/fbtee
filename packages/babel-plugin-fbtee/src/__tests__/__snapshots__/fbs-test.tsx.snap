// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test declarative (jsx) <fbs> syntax translation should convert a common string 1`] = `
import { fbs } from "fbtee";
const fbsCommonElem = fbs._(
  /* __FBT__ start */ {
    jsfbt: { m: [], t: { desc: "Button to post a comment", text: "Post" } },
    project: "",
  } /* __FBT__ end */
);

`;

exports[`Test declarative (jsx) <fbs> syntax translation should convert a simple string 1`] = `
import { fbs } from "fbtee";
const fbsElem = fbs._(
  /* __FBT__ start */ {
    jsfbt: { m: [], t: { desc: "str_description", text: "a simple string" } },
    project: "",
  } /* __FBT__ end */
);

`;

exports[`Test declarative (jsx) <fbs> syntax translation should convert a string with a parameter 1`] = `
import { fbs } from "fbtee";
const fbsElem = fbs._(
  /* __FBT__ start */ {
    jsfbt: {
      m: [],
      t: { desc: "str_description", text: "a string with a {param name}" },
    },
    project: "",
  } /* __FBT__ end */,
  [fbs._param("param name", parameter)]
);

`;

exports[`Test declarative (jsx) <fbs> syntax translation should handle <fbs:enum> 1`] = `
import { fbs } from "fbtee";
import aEnum from "Test$FbtEnum";
var x = fbs._(
  /* __FBT__ start */ {
    jsfbt: {
      m: [null],
      t: {
        id1: { desc: "enums!", text: "Click to see groups" },
        id2: { desc: "enums!", text: "Click to see photos" },
        id3: { desc: "enums!", text: "Click to see videos" },
      },
    },
    project: "",
  } /* __FBT__ end */,
  [fbs._enum(id, aEnum)]
);

`;

exports[`Test functional fbs() syntax translation should convert a common string 1`] = `
import { fbs } from "fbtee";
const fbsCommonCall = fbs._(
  /* __FBT__ start */ {
    jsfbt: { m: [], t: { desc: "Button to post a comment", text: "Post" } },
    project: "",
  } /* __FBT__ end */
);

`;

exports[`Test functional fbs() syntax translation should convert a simple string 1`] = `
import { fbs } from "fbtee";
const fbsCall = fbs._(
  /* __FBT__ start */ {
    jsfbt: { m: [], t: { desc: "str_description", text: "a simple string" } },
    project: "",
  } /* __FBT__ end */
);

`;

exports[`Test functional fbs() syntax translation should convert a string with a gender parameter 1`] = `
import { fbs } from "fbtee";
import IntlVariations from "IntlVariations";
const fbsCall = fbs._(
  /* __FBT__ start */ {
    jsfbt: {
      m: [{ token: "param name", type: 1 }],
      t: {
        "*": { desc: "str_description", text: "a string with a {param name}" },
      },
    },
    project: "",
  } /* __FBT__ end */,
  [fbs._param("param name", parameter, [1, IntlVariations.GENDER_MALE])]
);

`;

exports[`Test functional fbs() syntax translation should handle fbs.enum 1`] = `
import { fbs } from "fbtee";
import aEnum from "Test$FbtEnum";
var x = fbs._(
  /* __FBT__ start */ {
    jsfbt: {
      m: [null],
      t: {
        id1: { desc: "enums!", text: "Click to see groups" },
        id2: { desc: "enums!", text: "Click to see photos" },
        id3: { desc: "enums!", text: "Click to see videos" },
      },
    },
    project: "",
  } /* __FBT__ end */,
  [fbs._enum(id, aEnum)]
);

`;
